django/db/models/sql/query.py

```diff
@@ -419,6 +420,10 @@ def get_aggregation(self, using, aggregate_exprs):
                 getattr(self.annotations[ref], "subquery", False)
                 for ref in aggregate.get_refs()
             )
+            refs_window |= any(
```

Comment:
The following is a bit nitpicky, but: my first gut reaction in this code (because the diff was not showing the for-loop) was that `|=` was not needed. When I realized that this assignment was within the for-loop, I said "wow this is easy to miss, let's make sure the test covers this", and so I branched the code, made this change:

```diff
-            refs_window |= any(
+            refs_window = any(
```

and ran the tests. There was no test failure, do you think it'd be worth adding a test or extending the new one to cover this case?

(In all fairness, applying a similar change for `refs_subquery` does not produce any failure either :thinking: )

Reply:
It is needed because we're looping possibly multiple aggregations

[for alias, aggregate_expr in aggregate_exprs.items():](https://github.com/django/django/blob/da2f8e8257d1bea4215381684ca4abfcee333c43/django/db/models/sql/query.py#L408C9-L408C62)

If any of them refers to a window functions, just like it's the case with subqueries, the subquery wrapping *must* happen.

The code you are proposing would fail to pass in cases where an aggregate function, which doesn't refer to subqueries or window functions, is passed to `aggregate`.

Try the following changes locally to understand better

```diff
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index 1608d19480..d83a746304 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -420,7 +420,7 @@ def get_aggregation(self, using, aggregate_exprs):
                 getattr(self.annotations[ref], "subquery", False)
                 for ref in aggregate.get_refs()
             )
-            refs_window |= any(
+            refs_window = any(
                 getattr(self.annotations[ref], "contains_over_clause", True)
                 for ref in aggregate.get_refs()
             )
diff --git a/tests/aggregation/tests.py b/tests/aggregation/tests.py
index 05acf4e8db..acd397bb0f 100644
--- a/tests/aggregation/tests.py
+++ b/tests/aggregation/tests.py
@@ -2218,6 +2218,7 @@ def test_referenced_window_requires_wrapping(self):
         with self.assertNumQueries(1) as ctx:
             aggregate = total_books_qs.aggregate(
                 sum_avg_publisher_pages=Sum("avg_publisher_pages"),
+                cnt=Count("*"),
             )
         sql = ctx.captured_queries[0]["sql"].lower()
         self.assertEqual(sql.count("select"), 2, "Subquery wrapping required")
```

Basically `|=` instead of `=` translates to `refs_window = refs_window or any(...)` and ensures that the pushdown happens independently of the other in which aggregates are specified.

Reply:
Thanks @charettes for the explicit explanation, though perhaps my comment was not clear. I do know how `|=` work, my confusion was caused by initially looking at the GH diff (where the `for` is not seen without expanding the code) vs looking at the diff once I branched PR. In the latter case, the need for `|=` is clear and obvious.

The new diff you propose for the test (ie, adding `cnt=Count("*")`) is great because it makes the test fail if someone would naively remove the inplace OR operator, so I would suggest to add it to the provided test, plus the necessary adjustment in the following assertion:

```diff
-        self.assertEqual(aggregate, {"sum_avg_publisher_pages": 1100.0})
+        self.assertEqual(aggregate, {"sum_avg_publisher_pages": 1100.0, "cnt": 2})
```

Reply:
> is great because it makes the test fail if someone would naively remove the inplace OR operator,

I'm not convinced this is required tbh, there are multiple areas in the source that are not obvious where not looked at in their complete context. I would argue that reviewing code without fully expanding it might be the actual problem here? I mean, I don't think that we can realistically write code in a way that accounts for the fact that it might be reviewed out of context can we?

Reply:
Right, I understand your point, and I think we are in agreement. I personally like to branch the PRs and try things locally when reviewing.

Having said that, and despite I did reference the (lack of) diff context in my initial comment, my main point is about the `|=` usage (not) being tested. On retrospective, I should've been less verbose and more concise about this point since the start. Sorry about the ranting!